File: go/grainref/eval.go
package grainref

import (
	stdbytes "bytes"
	"math/bits"
)

// Eval evaluates formula against subject under the meter (spec §§1, 6–11).
// It returns (result, TrapNone) or (nil, trap). The reported charge is
// m.Spent() at return in both cases.
//
// The machine is fully iterative: an explicit continuation stack carries
// pending work, so host call-stack use never grows with evaluation depth
// (spec §13.1). Tail positions (final sub-evaluations of ops 2, 6, 7, 8, 9,
// 11 and cons) replace the current task without pushing a continuation.
func Eval(version uint32, subject, formula *Noun, m *Meter) (*Noun, Trap) {
	if version != GrainVersion {
		return nil, TrapUnknownVersion // charge 0, before any other observation
	}
	conts := make([]cont, 0, 32)
	s, f := subject, formula
	var v *Noun

EvalLoop:
	for {
		// ---- §8 dispatch of (s, f) ----
		// 1. atom formula
		if !f.IsCell() {
			return nil, TrapTypeMismatch // no charge
		}
		head, arg := f.head, f.tail

		// 3. cons composition: head is a cell
		if head.IsCell() {
			if t := m.charge(costCons); t != TrapNone {
				return nil, t
			}
			conts = append(conts, cont{kind: kConsFh, sub: s, f: arg})
			f = head
			continue EvalLoop
		}

		// 4. opcode range
		hb := head.bytes
		if len(hb) > 1 || (len(hb) == 1 && hb[0] > 11) {
			return nil, TrapUnknownOpcode // no charge
		}
		var op byte
		if len(hb) == 1 {
			op = hb[0]
		}

		// 5. shape validation (no charge), 6. dispatch charge, 7. reduction
		switch op {
		case 0: // slot
			if arg.IsCell() {
				return nil, TrapTypeMismatch
			}
			if len(arg.bytes) == 0 {
				return nil, TrapInvalidAxis
			}
			if t := chargeSlot(m, arg); t != TrapNone {
				return nil, t
			}
			r, t := walk(s, arg)
			if t != TrapNone {
				return nil, t
			}
			v = r // shared, no allocation

		case 1: // quote
			if t := m.charge(costQuote); t != TrapNone {
				return nil, t
			}
			v = arg // shared, no allocation

		case 2: // apply: *[s [2 b c]] = *[*[s b] *[s c]]
			if !arg.IsCell() {
				return nil, TrapTypeMismatch
			}
			if t := m.charge(costApply); t != TrapNone {
				return nil, t
			}
			conts = append(conts, cont{kind: kApplyB, sub: s, f: arg.tail})
			f = arg.head
			continue EvalLoop

		case 3: // is-cell
			if t := m.charge(costIsCell); t != TrapNone {
				return nil, t
			}
			conts = append(conts, cont{kind: kIsCell})
			f = arg
			continue EvalLoop

		case 4: // inc — no dispatch charge
			conts = append(conts, cont{kind: kInc})
			f = arg
			continue EvalLoop

		case 5: // equal — no dispatch charge
			if !arg.IsCell() {
				return nil, TrapTypeMismatch
			}
			conts = append(conts, cont{kind: kEqualB, sub: s, f: arg.tail})
			f = arg.head
			continue EvalLoop

		case 6: // if: arg must be [b [c d]]
			if !arg.IsCell() || !arg.tail.IsCell() {
				return nil, TrapTypeMismatch
			}
			if t := m.charge(costIf); t != TrapNone {
				return nil, t
			}
			conts = append(conts, cont{kind: kIf, sub: s, f: arg.tail.head, g: arg.tail.tail})
			f = arg.head
			continue EvalLoop

		case 7: // compose
			if !arg.IsCell() {
				return nil, TrapTypeMismatch
			}
			if t := m.charge(costCompose); t != TrapNone {
				return nil, t
			}
			conts = append(conts, cont{kind: kCompose, f: arg.tail})
			f = arg.head
			continue EvalLoop

		case 8: // push
			if !arg.IsCell() {
				return nil, TrapTypeMismatch
			}
			if t := m.charge(costPush); t != TrapNone {
				return nil, t
			}
			conts = append(conts, cont{kind: kPush, sub: s, f: arg.tail})
			f = arg.head
			continue EvalLoop

		case 9: // arm: arg [b c], b atom nonzero
			if !arg.IsCell() {
				return nil, TrapTypeMismatch
			}
			if arg.head.IsCell() {
				return nil, TrapTypeMismatch
			}
			if len(arg.head.bytes) == 0 {
				return nil, TrapInvalidAxis
			}
			if t := m.charge(costArm); t != TrapNone {
				return nil, t
			}
			conts = append(conts, cont{kind: kArm, ax: arg.head})
			f = arg.tail
			continue EvalLoop

		case 10: // edit: arg [[b c] d], b atom nonzero — no dispatch charge
			if !arg.IsCell() || !arg.head.IsCell() {
				return nil, TrapTypeMismatch
			}
			if arg.head.head.IsCell() {
				return nil, TrapTypeMismatch
			}
			if len(arg.head.head.bytes) == 0 {
				return nil, TrapInvalidAxis
			}
			conts = append(conts, cont{kind: kEditC, sub: s, ax: arg.head.head, f: arg.tail})
			f = arg.head.tail // evaluate c first, then d
			continue EvalLoop

		case 11: // hint: [11 h f] — h never evaluated, COST_HINT = 0
			if !arg.IsCell() {
				return nil, TrapTypeMismatch
			}
			f = arg.tail // tail position
			continue EvalLoop
		}

		// ---- return v through the continuation stack ----
	ReturnLoop:
		for {
			if len(conts) == 0 {
				return v, TrapNone
			}
			c := conts[len(conts)-1]
			conts = conts[:len(conts)-1]
			switch c.kind {
			case kConsFh: // have *[s fh]; evaluate ft
				conts = append(conts, cont{kind: kConsFt, x: v})
				s, f = c.sub, c.f
				continue EvalLoop
			case kConsFt: // have both halves; allocate the pair
				nc, t := allocCell(m, c.x, v)
				if t != TrapNone {
					return nil, t
				}
				v = nc
				continue ReturnLoop
			case kApplyB: // have the new subject; evaluate c
				conts = append(conts, cont{kind: kApplyC, x: v})
				s, f = c.sub, c.f
				continue EvalLoop
			case kApplyC: // have the new formula; tail-evaluate it
				s, f = c.x, v
				continue EvalLoop
			case kIsCell:
				if v.IsCell() {
					v = atomZero
				} else {
					v = atomOne
				}
				continue ReturnLoop
			case kInc:
				nv, t := incComplete(m, v)
				if t != TrapNone {
					return nil, t
				}
				v = nv
				continue ReturnLoop
			case kEqualB:
				conts = append(conts, cont{kind: kEqualC, x: v})
				s, f = c.sub, c.f
				continue EvalLoop
			case kEqualC:
				nv, t := equalComplete(m, c.x, v)
				if t != TrapNone {
					return nil, t
				}
				v = nv
				continue ReturnLoop
			case kIf: // v is the condition
				if v.IsCell() || len(v.bytes) > 1 {
					return nil, TrapTypeMismatch
				}
				if len(v.bytes) == 0 {
					f = c.f // atom 0: then-branch
				} else if v.bytes[0] == 1 {
					f = c.g // atom 1: else-branch
				} else {
					return nil, TrapTypeMismatch
				}
				s = c.sub
				continue EvalLoop
			case kCompose: // v is the new subject
				s, f = v, c.f
				continue EvalLoop
			case kPush: // v is the pushed value; allocate [v s]
				nc, t := allocCell(m, v, c.sub)
				if t != TrapNone {
					return nil, t
				}
				s, f = nc, c.f
				continue EvalLoop
			case kArm: // v is the core; slot charge, walk, tail-evaluate arm
				if t := chargeSlot(m, c.ax); t != TrapNone {
					return nil, t
				}
				af, t := walk(v, c.ax)
				if t != TrapNone {
					return nil, t
				}
				s, f = v, af
				continue EvalLoop
			case kEditC: // v is the replacement; evaluate d
				conts = append(conts, cont{kind: kEditD, ax: c.ax, x: v})
				s, f = c.sub, c.f
				continue EvalLoop
			case kEditD: // v is the target tree
				nv, t := editComplete(m, c.ax, c.x, v)
				if t != TrapNone {
					return nil, t
				}
				v = nv
				continue ReturnLoop
			}
		}
	}
}

type contKind uint8

const (
	kConsFh contKind = iota
	kConsFt
	kApplyB
	kApplyC
	kIsCell
	kInc
	kEqualB
	kEqualC
	kIf
	kCompose
	kPush
	kArm
	kEditC
	kEditD
)

// cont is one pending continuation frame of the iterative machine.
type cont struct {
	kind contKind
	sub  *Noun // saved subject
	x    *Noun // first computed value
	f    *Noun // pending sub-formula
	g    *Noun // second pending sub-formula (op 6 else-branch)
	ax   *Noun // axis atom (ops 9, 10)
}

// bitsOf returns the bit length of a nonzero atom (spec §7).
func bitsOf(a *Noun) uint64 {
	n := len(a.bytes)
	return uint64(n-1)*8 + uint64(bits.Len8(a.bytes[n-1]))
}

// chargeSlot charges the full slot cost for an axis before any walk:
// COST_SLOT_BASE + (bits(axis)-1) * COST_SLOT_STEP (spec §7).
func chargeSlot(m *Meter, ax *Noun) Trap {
	return m.charge(costSlotBase + (bitsOf(ax)-1)*costSlotStep)
}

// walk applies the §7 axis walk to n. The axis is a nonzero atom; the
// caller has already charged the slot cost where one is due.
func walk(n *Noun, ax *Noun) (*Noun, Trap) {
	for i := int64(bitsOf(ax)) - 2; i >= 0; i-- {
		if !n.IsCell() {
			return nil, TrapInvalidAxis
		}
		if ax.bytes[i>>3]>>(uint(i)&7)&1 == 0 {
			n = n.head
		} else {
			n = n.tail
		}
	}
	return n, TrapNone
}

// allocCell runs the §6 allocation sequence for one cell: charge 3 steps,
// add 3 arena words, check the depth bound, construct.
func allocCell(m *Meter, h, t *Noun) (*Noun, Trap) {
	if tr := m.charge(costCellAlloc); tr != TrapNone {
		return nil, tr
	}
	if tr := m.arenaAdd(cellWords); tr != TrapNone {
		return nil, tr
	}
	d := h.depth
	if t.depth > d {
		d = t.depth
	}
	if d+1 > MaxCellDepth {
		return nil, TrapNounOversized
	}
	return &Noun{head: h, tail: t, depth: d + 1}, TrapNone
}

// awords is the word size of an atom payload length (spec §2).
func awords(n uint64) uint64 { return (n + 7) / WordBytes }

// incComplete is the op 4 completion (spec §9): type check, operand charge,
// increment, ATOM_BOUND check, then the atom allocation sequence.
func incComplete(m *Meter, a *Noun) (*Noun, Trap) {
	if a.IsCell() {
		return nil, TrapTypeMismatch
	}
	n := uint64(len(a.bytes))
	if t := m.charge(costIncBase + awords(n)*costIncWord); t != TrapNone {
		return nil, t
	}
	r := make([]byte, n, n+1)
	copy(r, a.bytes)
	carry := true
	for i := 0; i < len(r) && carry; i++ {
		r[i]++
		carry = r[i] == 0
	}
	if carry {
		r = append(r, 1)
	}
	rlen := uint64(len(r))
	if rlen > MaxAtomBytes {
		return nil, TrapAtomBound
	}
	// Allocation sequence for an atom of rlen bytes.
	w := 1 + awords(rlen)
	if t := m.charge(w); t != TrapNone {
		return nil, t
	}
	if t := m.arenaAdd(w); t != TrapNone {
		return nil, t
	}
	return newAtom(r), TrapNone
}

// equalComplete is the op 5 completion: base charge, then the frozen
// comparison walk (spec §9). Physical-identity shortcuts must not change
// the charge, so the walk always runs in full until a verdict.
func equalComplete(m *Meter, x, y *Noun) (*Noun, Trap) {
	if t := m.charge(costEqualBase); t != TrapNone {
		return nil, t
	}
	type pair struct{ p, q *Noun }
	stack := make([]pair, 1, 32)
	stack[0] = pair{x, y}
	for len(stack) > 0 {
		pr := stack[len(stack)-1]
		stack = stack[:len(stack)-1]
		if t := m.charge(costEqualNode); t != TrapNone {
			return nil, t
		}
		p, q := pr.p, pr.q
		pc, qc := p.IsCell(), q.IsCell()
		switch {
		case !pc && !qc:
			if len(p.bytes) != len(q.bytes) {
				return atomOne, TrapNone // no word charge
			}
			if t := m.charge(awords(uint64(len(p.bytes))) * costEqualWord); t != TrapNone {
				return nil, t
			}
			if !stdbytes.Equal(p.bytes, q.bytes) {
				return atomOne, TrapNone
			}
		case pc && qc:
			stack = append(stack, pair{p.tail, q.tail}, pair{p.head, q.head}) // heads compared first
		default:
			return atomOne, TrapNone
		}
	}
	return atomZero, TrapNone
}

// editComplete is the op 10 completion (spec §9), in frozen order:
// 1. charge COST_EDIT_BASE + L*COST_EDIT_STEP (L = bits(axis)-1);
// 2. walk t along the axis path (INVALID_AXIS on atom descent);
// 3. arena_add(3*L);
// 4. rebuild the spine bottom-up with a depth check per new cell.
// The completion charge pre-pays the per-level allocation steps; step 3
// adds only the words.
func editComplete(m *Meter, ax *Noun, v, t *Noun) (*Noun, Trap) {
	l := bitsOf(ax) - 1
	if tr := m.charge(costEditBase + l*costEditStep); tr != TrapNone {
		return nil, tr
	}
	if l == 0 {
		return v, TrapNone // #[1, v, t] = v: no walk, no allocation
	}
	// Walk, recording each level's direction bit and untouched sibling.
	dirs := make([]byte, 0, l)
	sibs := make([]*Noun, 0, l)
	cur := t
	for i := int64(l) - 1; i >= 0; i-- {
		if !cur.IsCell() {
			return nil, TrapInvalidAxis
		}
		b := ax.bytes[i>>3] >> (uint(i) & 7) & 1
		dirs = append(dirs, b)
		if b == 0 {
			sibs = append(sibs, cur.tail)
			cur = cur.head
		} else {
			sibs = append(sibs, cur.head)
			cur = cur.tail
		}
	}
	if tr := m.arenaAdd(3 * l); tr != TrapNone {
		return nil, tr
	}
	// Rebuild bottom-up: L new cells, each reusing the untouched sibling.
	res := v
	for j := len(dirs) - 1; j >= 0; j-- {
		var h, tl *Noun
		if dirs[j] == 0 {
			h, tl = res, sibs[j]
		} else {
			h, tl = sibs[j], res
		}
		d := h.depth
		if tl.depth > d {
			d = tl.depth
		}
		if d+1 > MaxCellDepth {
			return nil, TrapNounOversized
		}
		res = &Noun{head: h, tail: tl, depth: d + 1}
	}
	return res, TrapNone
}

File: crates/noos-grain/src/eval.rs
//! The Grain v1 machine (spec §§7–10): explicit work-stack interpreter, no
//! host recursion, charge-before-reduction/allocation, deterministic traps.

use crate::noun::Noun;
use crate::{
    awords, GrainTrap, Meter, COST_APPLY, COST_ARM, COST_CELL_ALLOC, COST_COMPOSE, COST_CONS,
    COST_EDIT_BASE, COST_EDIT_STEP, COST_EQUAL_BASE, COST_EQUAL_NODE, COST_EQUAL_WORD, COST_IF,
    COST_INC_BASE, COST_INC_WORD, COST_ISCELL, COST_PUSH, COST_QUOTE, COST_SLOT_BASE,
    COST_SLOT_STEP, GRAIN_VERSION, MAX_ATOM_BYTES,
};

enum Task {
    Eval(Noun, Noun),
    Kont(K),
}

enum K {
    ConsMake,
    Apply,
    IsCell,
    Inc,
    Equal,
    If {
        subject: Noun,
        then_f: Noun,
        else_f: Noun,
    },
    Compose {
        formula: Noun,
    },
    Push {
        subject: Noun,
        formula: Noun,
    },
    Arm {
        axis: Noun,
    },
    Edit {
        axis: Noun,
    },
}

/// Lab-only jet hook surface. [`eval_with_jets`] recognizes `[12 id f]` and
/// offers the evaluation to the hook. A firing hook MUST reproduce the exact
/// observational triple of evaluating `f` on the subject — value-or-trap,
/// trap code, and metering — charging `meter` through the same frozen
/// schedule. Declining (`None`) falls back to interpreting `f`, so erasing
/// `[12 id f]` to `f` is always semantics-preserving, exactly like opcode
/// 11. Production [`eval`] passes no hook: opcode 12 stays `UNKNOWN_OPCODE`.
pub trait JetHook {
    /// `id` is pure data (never evaluated); return `None` to decline.
    fn dispatch(
        &self,
        id: &Noun,
        subject: &Noun,
        formula: &Noun,
        meter: &mut Meter,
    ) -> Option<Result<Noun, GrainTrap>>;
}

/// Pure deterministic evaluation (spec §1). The conformance triple is
/// `(value-or-trap, trap_code, meter.spent())`.
pub fn eval(
    version: u32,
    subject: Noun,
    formula: Noun,
    meter: &mut Meter,
) -> Result<Noun, GrainTrap> {
    eval_inner(version, subject, formula, meter, None)
}

/// [`eval`] with a lab-only jet hook: identical semantics except that
/// `[12 id f]` consults `hook` (see [`JetHook`]).
pub fn eval_with_jets(
    version: u32,
    subject: Noun,
    formula: Noun,
    meter: &mut Meter,
    hook: &dyn JetHook,
) -> Result<Noun, GrainTrap> {
    eval_inner(version, subject, formula, meter, Some(hook))
}

fn eval_inner(
    version: u32,
    subject: Noun,
    formula: Noun,
    meter: &mut Meter,
    hook: Option<&dyn JetHook>,
) -> Result<Noun, GrainTrap> {
    if version != GRAIN_VERSION {
        return Err(GrainTrap::UnknownVersion);
    }
    let mut tasks: Vec<Task> = vec![Task::Eval(subject, formula)];
    let mut vals: Vec<Noun> = Vec::new();

    while let Some(task) = tasks.pop() {
        match task {
            Task::Eval(s, f) => dispatch(s, f, &mut tasks, &mut vals, meter, hook)?,
            Task::Kont(k) => kont(k, &mut tasks, &mut vals, meter)?,
        }
    }

    debug_assert_eq!(vals.len(), 1);
    // Unreachable fallback: the machine pushes exactly one value per Eval.
    vals.pop().ok_or(GrainTrap::TypeMismatch)
}

/// Spec §8: shape validation (uncharged) → dispatch charge → schedule.
fn dispatch(
    s: Noun,
    f: Noun,
    tasks: &mut Vec<Task>,
    vals: &mut Vec<Noun>,
    meter: &mut Meter,
    hook: Option<&dyn JetHook>,
) -> Result<(), GrainTrap> {
    let (head, arg) = match f.as_cell() {
        Some((h, t)) => (h.clone(), t.clone()),
        None => return Err(GrainTrap::TypeMismatch),
    };

    // Cons composition: head is itself a formula cell.
    if head.is_cell() {
        meter.charge(COST_CONS)?;
        tasks.push(Task::Kont(K::ConsMake));
        tasks.push(Task::Eval(s.clone(), arg)); // tail formula, second
        tasks.push(Task::Eval(s, head)); // head formula, first
        return Ok(());
    }

    let op = match head.as_atom() {
        Some([]) => 0u8,
        Some(&[b]) if b <= 12 => b,
        _ => return Err(GrainTrap::UnknownOpcode),
    };

    match op {
        0 => {
            let axis = arg;
            check_axis(&axis)?;
            meter.charge(slot_cost(&axis))?;
            let r = walk(&s, &axis)?;
            vals.push(r);
        }
        1 => {
            meter.charge(COST_QUOTE)?;
            vals.push(arg);
        }
        2 => {
            let (b, c) = split(&arg)?;
            meter.charge(COST_APPLY)?;
            tasks.push(Task::Kont(K::Apply));
            tasks.push(Task::Eval(s.clone(), c)); // new formula, second
            tasks.push(Task::Eval(s, b)); // new subject, first
        }
        3 => {
            meter.charge(COST_ISCELL)?;
            tasks.push(Task::Kont(K::IsCell));
            tasks.push(Task::Eval(s, arg));
        }
        4 => {
            // All charges at completion (operand-size dependent).
            tasks.push(Task::Kont(K::Inc));
            tasks.push(Task::Eval(s, arg));
        }
        5 => {
            let (b, c) = split(&arg)?;
            tasks.push(Task::Kont(K::Equal));
            tasks.push(Task::Eval(s.clone(), c));
            tasks.push(Task::Eval(s, b));
        }
        6 => {
            let (b, cd) = split(&arg)?;
            let (c, d) = split(&cd)?;
            meter.charge(COST_IF)?;
            tasks.push(Task::Kont(K::If {
                subject: s.clone(),
                then_f: c,
                else_f: d,
            }));
            tasks.push(Task::Eval(s, b));
        }
        7 => {
            let (b, c) = split(&arg)?;
            meter.charge(COST_COMPOSE)?;
            tasks.push(Task::Kont(K::Compose { formula: c }));
            tasks.push(Task::Eval(s, b));
        }
        8 => {
            let (b, c) = split(&arg)?;
            meter.charge(COST_PUSH)?;
            tasks.push(Task::Kont(K::Push {
                subject: s.clone(),
                formula: c,
            }));
            tasks.push(Task::Eval(s, b));
        }
        9 => {
            let (b, c) = split(&arg)?;
            if b.is_cell() {
                return Err(GrainTrap::TypeMismatch);
            }
            check_axis(&b)?;
            meter.charge(COST_ARM)?;
            tasks.push(Task::Kont(K::Arm { axis: b }));
            tasks.push(Task::Eval(s, c));
        }
        10 => {
            let (bc, d) = split(&arg)?;
            let (b, c) = split(&bc)?;
            if b.is_cell() {
                return Err(GrainTrap::TypeMismatch);
            }
            check_axis(&b)?;
            // All charges at completion (axis-length dependent).
            tasks.push(Task::Kont(K::Edit { axis: b }));
            tasks.push(Task::Eval(s.clone(), d)); // old tree, second
            tasks.push(Task::Eval(s, c)); // replacement value, first
        }
        11 => {
            // [11 h f]: h is pure data, never evaluated; COST_HINT = 0.
            let (_hint, f2) = split(&arg)?;
            tasks.push(Task::Eval(s, f2));
        }
        12 => {
            // Lab-only jet hint: invalid in production (no hook), checked
            // BEFORE any shape validation so production behavior for every
            // `[12 ...]` formula is exactly `UNKNOWN_OPCODE`. With a hook,
            // `[12 id f]` mirrors opcode 11 (charge 0, `id` never
            // evaluated): the hook may produce the exact observational
            // triple of `f`, else `f` is interpreted.
            let Some(h) = hook else {
                return Err(GrainTrap::UnknownOpcode);
            };
            let (id, f2) = split(&arg)?;
            match h.dispatch(&id, &s, &f2, meter) {
                Some(r) => vals.push(r?),
                None => tasks.push(Task::Eval(s, f2)),
            }
        }
        _ => return Err(GrainTrap::UnknownOpcode), // unreachable: op <= 12
    }
    Ok(())
}

fn kont(
    k: K,
    tasks: &mut Vec<Task>,
    vals: &mut Vec<Noun>,
    meter: &mut Meter,
) -> Result<(), GrainTrap> {
    match k {
        K::ConsMake => {
            let tail = pop(vals);
            let head = pop(vals);
            let cell = alloc_cell(head, tail, meter)?;
            vals.push(cell);
        }
        K::Apply => {
            let f2 = pop(vals);
            let s2 = pop(vals);
            tasks.push(Task::Eval(s2, f2));
        }
        K::IsCell => {
            let v = pop(vals);
            vals.push(loobean(v.is_cell()));
        }
        K::Inc => {
            let a = pop(vals);
            let bytes = a.as_atom().ok_or(GrainTrap::TypeMismatch)?;
            meter.charge(
                COST_INC_BASE.saturating_add(awords(bytes.len()).saturating_mul(COST_INC_WORD)),
            )?;
            let mut r = bytes.to_vec();
            let mut carried = true;
            for b in r.iter_mut() {
                let (nb, overflow) = b.overflowing_add(1);
                *b = nb;
                if !overflow {
                    carried = false;
                    break;
                }
            }
            if carried {
                if r.len() >= MAX_ATOM_BYTES {
                    return Err(GrainTrap::AtomBound);
                }
                r.push(1);
            }
            let out = alloc_atom(r, meter)?;
            vals.push(out);
        }
        K::Equal => {
            let y = pop(vals);
            let x = pop(vals);
            meter.charge(COST_EQUAL_BASE)?;
            let eq = metered_eq(&x, &y, meter)?;
            vals.push(loobean(eq));
        }
        K::If {
            subject,
            then_f,
            else_f,
        } => {
            let cond = pop(vals);
            let branch = match cond.as_atom() {
                Some([]) => then_f,
                Some([1]) => else_f,
                _ => return Err(GrainTrap::TypeMismatch),
            };
            tasks.push(Task::Eval(subject, branch));
        }
        K::Compose { formula } => {
            let s2 = pop(vals);
            tasks.push(Task::Eval(s2, formula));
        }
        K::Push { subject, formula } => {
            let v = pop(vals);
            let s2 = alloc_cell(v, subject, meter)?;
            tasks.push(Task::Eval(s2, formula));
        }
        K::Arm { axis } => {
            let core = pop(vals);
            meter.charge(slot_cost(&axis))?;
            let f = walk(&core, &axis)?;
            tasks.push(Task::Eval(core, f));
        }
        K::Edit { axis } => {
            let t = pop(vals);
            let v = pop(vals);
            let out = edit(&axis, v, t, meter)?;
            vals.push(out);
        }
    }
    Ok(())
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

/// Machine invariant: every Kont finds its operands on the value stack.
#[inline]
fn pop(vals: &mut Vec<Noun>) -> Noun {
    match vals.pop() {
        Some(v) => v,
        None => unreachable!("value stack underflow"),
    }
}

#[inline]
fn split(n: &Noun) -> Result<(Noun, Noun), GrainTrap> {
    match n.as_cell() {
        Some((h, t)) => Ok((h.clone(), t.clone())),
        None => Err(GrainTrap::TypeMismatch),
    }
}

/// Loobean result constants: canonical atoms, no allocation charge (spec §6).
#[inline]
fn loobean(yes: bool) -> Noun {
    if yes {
        Noun::atom_u64(0)
    } else {
        Noun::atom_u64(1)
    }
}

/// Axis operand must be an atom (TYPE_MISMATCH) and nonzero (INVALID_AXIS);
/// both checked before any charge (spec §8 step 5).
fn check_axis(axis: &Noun) -> Result<(), GrainTrap> {
    match axis.as_atom() {
        None => Err(GrainTrap::TypeMismatch),
        Some([]) => Err(GrainTrap::InvalidAxis),
        Some(_) => Ok(()),
    }
}

/// Bit length of a nonzero minimal-LE atom.
fn bits(bytes: &[u8]) -> u64 {
    debug_assert!(!bytes.is_empty());
    let top = *bytes.last().unwrap_or(&0);
    debug_assert!(top != 0);
    // u8::leading_zeros counts within 8 bits.
    ((bytes.len() as u64).saturating_sub(1))
        .saturating_mul(8)
        .saturating_add(8u64.saturating_sub(u64::from(top.leading_zeros())))
}

#[inline]
fn axis_bit(bytes: &[u8], i: u64) -> bool {
    let byte = bytes[(i / 8) as usize];
    (byte >> (i % 8)) & 1 == 1
}

/// Slot cost (spec §7): `COST_SLOT_BASE + (bits(axis) - 1) * COST_SLOT_STEP`.
fn slot_cost(axis: &Noun) -> u64 {
    let b = match axis.as_atom() {
        Some(bytes) => bits(bytes),
        None => unreachable!("slot_cost on a cell axis"),
    };
    COST_SLOT_BASE.saturating_add(b.saturating_sub(1).saturating_mul(COST_SLOT_STEP))
}

/// Axis walk (spec §7). The caller has already charged the slot cost.
fn walk(n: &Noun, axis: &Noun) -> Result<Noun, GrainTrap> {
    let bytes = axis.as_atom().ok_or(GrainTrap::TypeMismatch)?;
    if bytes.is_empty() {
        return Err(GrainTrap::InvalidAxis);
    }
    let nbits = bits(bytes);
    let mut cur = n.clone();
    let mut i = nbits.saturating_sub(1);
    while i > 0 {
        i = i.saturating_sub(1);
        let (h, t) = match cur.as_cell() {
            Some((h, t)) => (h.clone(), t.clone()),
            None => return Err(GrainTrap::InvalidAxis),
        };
        cur = if axis_bit(bytes, i) { t } else { h };
    }
    Ok(cur)
}

/// Allocation sequence (spec §6): charge → arena → depth check → construct.
fn alloc_cell(head: Noun, tail: Noun, meter: &mut Meter) -> Result<Noun, GrainTrap> {
    meter.charge(COST_CELL_ALLOC)?;
    meter.arena_add(COST_CELL_ALLOC)?;
    Noun::cell(head, tail)
}

/// Allocation sequence for an atom of minimal bytes `r`.
fn alloc_atom(r: Vec<u8>, meter: &mut Meter) -> Result<Noun, GrainTrap> {
    let w = 1u64.saturating_add(awords(r.len()));
    meter.charge(w)?;
    meter.arena_add(w)?;
    Ok(Noun::atom_from_minimal(r))
}

/// Opcode 5 comparison walk with the frozen charge schedule (spec §9 op 5).
fn metered_eq(x: &Noun, y: &Noun, meter: &mut Meter) -> Result<bool, GrainTrap> {
    let mut stack: Vec<(Noun, Noun)> = vec![(x.clone(), y.clone())];
    while let Some((p, q)) = stack.pop() {
        meter.charge(COST_EQUAL_NODE)?;
        match (p.as_atom(), q.as_atom()) {
            (Some(a), Some(b)) => {
                if a.len() != b.len() {
                    return Ok(false);
                }
                meter.charge(awords(a.len()).saturating_mul(COST_EQUAL_WORD))?;
                if a != b {
                    return Ok(false);
                }
            }
            (None, None) => {
                let (ph, pt) = split(&p)?;
                let (qh, qt) = split(&q)?;
                stack.push((pt, qt));
                stack.push((ph, qh)); // heads compared first
            }
            _ => return Ok(false),
        }
    }
    Ok(true)
}

/// Opcode 10 completion (spec §9 op 10).
fn edit(axis: &Noun, v: Noun, t: Noun, meter: &mut Meter) -> Result<Noun, GrainTrap> {
    let bytes = axis.as_atom().ok_or(GrainTrap::TypeMismatch)?;
    if bytes.is_empty() {
        return Err(GrainTrap::InvalidAxis);
    }
    let levels = bits(bytes).saturating_sub(1);
    // 1. Bulk charge: base + per-level (walk step + that level's cell).
    meter.charge(COST_EDIT_BASE.saturating_add(levels.saturating_mul(COST_EDIT_STEP)))?;
    // 2. Walk, recording (sibling, went_tail) per level.
    let mut path: Vec<(Noun, bool)> = Vec::with_capacity(levels as usize);
    let mut cur = t;
    let mut i = levels;
    while i > 0 {
        i = i.saturating_sub(1);
        let (h, tl) = match cur.as_cell() {
            Some((h, tl)) => (h.clone(), tl.clone()),
            None => return Err(GrainTrap::InvalidAxis),
        };
        if axis_bit(bytes, i) {
            path.push((h, true));
            cur = tl;
        } else {
            path.push((tl, false));
            cur = h;
        }
    }
    // 3. Arena words for the rebuilt spine.
    meter.arena_add(levels.saturating_mul(COST_CELL_ALLOC))?;
    // 4. Rebuild bottom-up; each new cell's depth is checked.
    let mut acc = v;
    for (sibling, went_tail) in path.into_iter().rev() {
        acc = if went_tail {
            Noun::cell(sibling, acc)?
        } else {
            Noun::cell(acc, sibling)?
        };
    }
    Ok(acc)
}
